code/modules/reagents/addiction/_addiction.dm

```diff
@@ -0,0 +1,118 @@
+/// Base class for addiction, handles when you become addicted and what the effects of that are. By default you become addicted when you hit a certain threshold, and stop being addicted once you go below another one.
+/datum/addiction
+	/// Name of this addiction
+	var/name = "Unknown Addiction"
+	/// Higher threshold, when you start being addicted
+	var/addiction_gain_threshold = 600
+	/// Lower threshold, when you stop being addicted
+	var/addiction_loss_threshold = 400
+	/// Messages for each stage of addictions.
+	var/list/withdrawal_stage_messages = list()
+	/// Rates at which you lose addiction (in units/second) if you are not on the drug at that time per stage
+	var/addiction_loss_per_stage = list(0.5, 0.5, 1, 1.5)
+	/// Amount of drugs you need in your system to be satisfied
+	var/addiction_relief_treshold = MIN_ADDICTION_REAGENT_AMOUNT
+
+/// Called when you gain addiction points somehow. Takes a carbon mob as argument and sees if you gained the addiction
+/datum/addiction/proc/OnGainAddictionPoints(mob/living/carbon/victim)
+	var/current_addiction_point_amount = victim.addiction_points[type]
+	if(current_addiction_point_amount < addiction_gain_threshold) //Not enough to become addicted
+		return
+	if(LAZYACCESS(victim.active_addictions, type)) //Already addicted
+		return
+	BecomeAddicted(victim)
+
+///Called when you become addicted
+/datum/addiction/proc/BecomeAddicted(mob/living/carbon/victim)
+	LAZYSET(victim.active_addictions, type, 1) // Start at first cycle.
+	log_game("[key_name(victim)] has become addicted to [name].")
+
+/// Called when you lose addiction points somehow. Takes a mind as argument and sees if you lost the addiction
+/datum/addiction/proc/OnLoseAddictionPoints(mob/living/carbon/victim)
+	var/current_addiction_point_amount = victim.addiction_points[type]
+	if(!LAZYACCESS(victim.active_addictions, type)) // Not addicted
+		return FALSE
+	if(current_addiction_point_amount > addiction_loss_threshold) // Not enough to stop being addicted
+		return FALSE
+	LoseAddiction(victim)
+	return TRUE
+
+/datum/addiction/proc/LoseAddiction(mob/living/carbon/victim)
+	to_chat(victim, SPAN_NOTICE("You feel like you've gotten over your need for drugs."))
```

Comment:
This needs to be more specific for addictions or at least take addiction name...

Reply:
I'd imagine you wouldn't immediately know that you are no longer alcohol addict if you got over it, but I considered such a thing.

Reply:
Then make it more vague instead of "drugs"

Reply:
Going to just remove the message altogether.